# Перед тем, как начать

Напоминаем вам, что блокнот - это интерактивная среда. В тексте блокнота встречаются слова **"попробуйте"**, **"запустите"** и так далее. Это значит, что вы нажимаете на ячейку кода и жмете ctrl + enter или кнопку "запустить / run"

---

### Также в блокноте есть **упражнения**. В них вам необходимо самостоятельно на основе теории написать код решения.

---

# В предыдущих сериях
На прошлом занятии мы познакомились с переменными. Переменная имеет имя и значение. Имя переменной может начинаться только с буквы и включать в себя буквы, цифры и символ подчеркивания. Имя переменной должно отражать ее назначение.

Чтобы задать переменной значение, необходимо после знака равно (оператора присваивания) указать значение переменной.

Еще значение переменной можно получить из ввода. Для этого используем команду input(). В этом случае значение переменной задает пользователь.

Также можно вывести значение переменной в консоль. Для этого нужно воспользоваться командой print(). В нее можно передать как непосредственно значение в виде 5 или 'Привет', так и переменные

```python
hello = 'Привет, Вася!'
print(hello)
```

Получая на выходе

```
Привет, Вася!
```

# Минутка полезности

## Унарные операторы
Это операторы, имеющие один операнд

Например, в -5 знак минуса - это унарный оператор, так как знак минуса **не требует** в данном случае **наличия второго числа**. А для случая 5 - 2 минус - это бинарный оператор, так как требуется **два** (**би**)  оператора.

# Условный оператор

>Условный оператор используется, когда некая часть программы должна быть выполнена, только если верно какое-либо условие. Для записи условного оператора используются ключевые слова if и else («если», «иначе»), двоеточие и  отступ в четыре пробела. В современных средах разработки (и в **Google Colab**) при нажатии **tab** происходит автоматическая замена символа табуляции на 4 пробела. В этом вы можете убедиться, открыв файл блокнота в текстовом редакторе.

```python
if условие:
    Действия, если условие верно
else:
    Действия, если условие неверно
```

>**Важно!** Python 3 запрещает смешивание табуляции и пробелов в отступах.

![title](images/1.jpg)

Рассмотрим пример. **Запустите код**. Обращаю ваше **внимание** на то, что программы с функцией **input** ожидают от вас ввода. После запуска вы увидите окно ввода под ячейкой. Туда и необходимо произвести ввод с клавиатуры:

In [5]:
print('Введите пароль:')
password = input()
if password == 'qwerty':
    print('Доступ открыт.')
else:
    print('Ошибка, доступ закрыт!')

Введите пароль:
Доступ открыт.


> В отличие от предыдущего занятия, тут появления **ветвление**. Мы записываем **логическое выражение** password == 'qwerty', которое может иметь **два исхода**. Либо доступ будет открыт, либо - закрыт.
> В первом занятии по Python мы писали другой тип алгоритмов, который называется **линейным**. В дальнейшем мы рассмотрим также **циклический тип** алгоритмов, однако пока сосредоточимся на ветвлении.

Обратите внимание: в начале условного оператора if выполняется сравнение, а не присваивание. Разница представлена ниже.

## Сравнение

Сравнение — это проверка, которая не меняет значение переменной (в сравнении может вообще не быть переменных), а присваивание — команда, которая меняет значение переменной.

Для сравнения нужно использовать двойной знак равенства: ==.

Также заметьте, что после else никогда не пишется никакого условия.

Еще пример. **Запустите**:

In [6]:
print('Представься, о незнакомец!')
name = input()
if name == 'Цезарь':
    print('Аве, Цезарь!')
else:
    print('Приветик.')

Представься, о незнакомец!
Аве, Цезарь!


В качестве условия можно использовать и другие операции отношения:

`<` меньше

`>` больше

`<=` меньше или равно

`>=` больше или равно

`==` равно

`!=` не равно

Обратите внимание, что операция сравнения == имеет два знака равно. Это сделано для того, чтобы не возникало конфликта с оператором присваивания =. Ниже представлен код, демонстрирующий применение этих двух операторов

In [61]:
5 == 5

True

В примере выше логическое выражение дает нам результата `True`. В питоне логические выражения всегда дают нам либо `True`, либо `False`. Да, нет. 1, 0. Они реализуют своего рода закрытые вопросы, на которые можно дать либо положительный, либо отрицательный ответ. Никогда между.

In [60]:
5 = 5

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (3619573732.py, line 1)

In [59]:
a = 5
print(a)

5


---
Помимо того, что логическое выражение имеет значение, это значение можно присваивать переменным

In [58]:
a = 1 == 2
print(a)

False


### PEP 8

Все операции отношения оформляются с помощью симметричных пробелов.

**Правильно:**


```python if bird == "Тук-тук": ```

**~~Неправильно:~~**

```if bird=="Тук-тук":```

Объекты любой однородной группы можно сравнивать между собой. Очевидно, что сравнение чисел происходит так, как мы привыкли это делать - 5 больше 3, -2 меньше 0. Подумайте над тем, как можно сравнивать, например, строки.

## Упражнения

Напишите программу, которая считывает одну строку. Если это строка «Python», программа выводит «ДА»; в противном случае программа выводит «НЕТ».

In [12]:
if input("Введите строку: ") == "Python":
    print("ДА")
else:
    print("НЕТ")  

ДА


Напишите программу, которая считывает одну строку. Если это пустая строка, т. е. если пользователь сразу нажал Enter, то программа выводит «ДА»; в противном случае программа выводит «НЕТ» 

In [14]:
if input("Введите строку: ") == "":
    print("ДА") 
else:
    print("НЕТ")

ДА


## Сложное условие. Логические операции

Иногда в условном операторе нужно задать сложное условие. Для этого можно использовать логические операции **and** («и»), **or** («или») и **not** («не»).

>Чтобы задать одновременное выполнение двух условий, используем and («и»), если достаточно выполнения одного из двух вариантов (или обоих сразу) — используем or («или»), а если нужно убрать какой-то вариант — not («не»).

**Приоритет выполнения операций**. Порядок важности (приоритета) сохранен, not выполняется перед and, and - перед or:

1. not
2. and
3. or

Если нужно изменить приоритет операций или вы забыли правила, используйте скобки.

Например, вот так можно проверить, что оба условия выполнены. **Запустите**. Пробуйте разные комбинации ввода, чтобы лучше понять код

1. альфа, омега
1. альфа, гриб
1. ленин, омега
1. бомба, водород

In [20]:
print('Как называются первая и последняя буквы греческого алфавита?')
greek_letter_1 = input()
greek_letter_2 = input()
if greek_letter_1 == 'альфа' and greek_letter_2 == 'омега':
    print('Верно.')
else:
    print('Неверно.')

Как называются первая и последняя буквы греческого алфавита?
Верно.


Ниже еще несколько примеров. **Посмотрите**, что дает их запуск.

In [23]:
print('Как греки или римляне называли главу своего пантеона - бога грома?')
ancient_god = input()
if ancient_god == 'Зевс' or ancient_god == 'Юпитер':
    print('Верно.')
else:
    print('Неверно.')

Как греки или римляне называли главу своего пантеона - бога грома?
Верно.


В данном случае сделайте так, чтобы ячейка ниже вывела слово "Верно"

In [24]:
print('Введите имена двух братьев из античных мифов и легенд.')
brother1 = input()
brother2 = input()
if brother1 == 'Ромул' and brother2 == 'Рем' or brother1 == 'Кастор' and (brother2 == 'Поллукс' or brother2 == 'Полидевк'):
    print('Верно.')
else:
    print('Неверно.')

Введите имена двух братьев из античных мифов и легенд.
Верно.


В ячейке выше вы можете заметить, что код на 4 строке может выходить за правый край ячейки кода в зависимости от разрешения экрана. Поэтому стоит напомнить

## PEP 8

По стандарту PEP 8 длина строки должна быть ограничена максимум 79 символами.

Есть несколько способов переноса длинных строк.

- Использование подразумеваемых продолжений Python внутри круглых, квадратных и фигурных скобок: длинные строки могут быть разбиты на несколько строк, заключенных в скобки.
- Использование символа ’\’ (обратный слеш, или бэкслеш) для обозначения места разрыва строки.

Мы будем использовать первый способ.

Если после перенесенной строки идет один или несколько вложенных операторов (например, мы переносим строку с условием в операторе if), отступ у перенесенной части должен быть на четыре пробела больше, чем у вложенного оператора.

Тогда представленный выше программный код может быть записан так:

In [57]:
print('Введите имена двух братьев из античных мифов и легенд.')
brother1 = input()
brother2 = input()
if (brother1 == 'Ромул' and brother2 == 'Рем' or brother1 == 'Кастор'
        and (brother2 == 'Поллукс' or brother2 == 'Полидевк')):
    print('Верно.')
else:
    print('Неверно.')

Введите имена двух братьев из античных мифов и легенд.
Верно.


Рассмотрим еще несколько примеров. **Запустите**

In [25]:
print('Введите любые два слова, но это не должны быть "белый" и "медведь" разом.')
word1 = input()
word2 = input()
if not (word1 == 'белый' and word2 == 'медведь'):
    print('Верно.')
else:
    print('Неверно.')

Введите любые два слова, но это не должны быть "белый" и "медведь" разом.
Верно.


## Упражнения

Напишите программу, которая считывает две строки и выводит «ВЕРНО», если в каждой из них записано или слово да, или слово нет (в любой комбинации). Если это не так, выведите «НЕВЕРНО». 

Комбинации для ВЕРНО. Остальные неверны

1. да, да
2. да, нет
3. нет, да
4. нет, нет

In [56]:
a = input()
b = input()

if (a == 'да' or a == 'нет') and (b == 'да' or b == 'нет'):
    print('ВЕРНО')

else:
    print('НЕВЕРНО')

ВЕРНО


Напишите программу, которая считывает три строки. Если эти три строки – «раз», «два» и «три», то программа выводит «ГОРИ», если нет, то «НЕ ГОРИ».

Обратите внимание на то, что порядок слов важен. Сначала раз, потом два, потом три.

In [27]:
# код
if input("Введите первую строку: ") == "раз" and \
   input("Введите вторую строку: ") == "два" and \
   input("Введите третью строку: ") == "три":
    print("ГОРИ")
else:
    print("НЕ ГОРИ") 

НЕ ГОРИ


Усовершенствуйте предыдущую программу так, чтобы не только при вводе «раз», «два», «три», но и при вводе «1», «2» и «3» тоже выводилось «ГОРИ». Смешанный ввод (например, «1», «2», «три») даёт «НЕ ГОРИ».

In [28]:
# код
input1 = input("Введите первую строку: ")
input2 = input("Введите вторую строку: ")
input3 = input("Введите третью строку: ")
if (input1 == "раз" and input2 == "два" and input3 == "три") or \
   (input1 == "1" and input2 == "2" and input3 == "3"):
    print("ГОРИ") 
else:
    print("НЕ ГОРИ")

ГОРИ


Добавьте в предыдущую программу возможность вместо «раз» ввести «один».

In [29]:
# код
input1 = input("Введите первую строку: ")
input2 = input("Введите вторую строку: ")
input3 = input("Введите третью строку: ")
if (input1 == ["раз", "один"] and input2 == "два" and input3 == "три") or \
   (input1 == "1" and input2 == "2" and input3 == "3"):
    print("ГОРИ") 
else:
    print("НЕ ГОРИ")

НЕ ГОРИ


# Вложенные условия

### Блок (секция) кода

В команде if при выполнении условия можно **выполнять более одной команды**. Для этого все их необходимо выделить отступом. Такая запись называется блоком кода. По отступам интерпретатор определяет, при выполнении каких условий какие команды исполнять. Аналогично можно делать и для команды else.

**Далее мы не будем напоминать вам о том, что код в ячейках необходимо запустить. Видите код - запускайте**.

Посмотрим, насколько вы превосходите субъект исследования товарища Павлова

>И **обязательно** пробуйте вводить разные значения. Смотрите на код, пытайтесь предсказать, какие значения дадут какие результаты.

In [32]:
print('Представься, о незнакомец!')
name = input()
if name == 'Цезарь' or name == 'Caesar':
    print('Аве, Цезарь!')
    print('Слава императору!')
else:
    print('Приветик.')
    print('Погода сегодня хорошая.')
print('Засим - заканчиваем.')

Представься, о незнакомец!
Приветик.
Погода сегодня хорошая.
Засим - заканчиваем.


Перед последней строчкой нет отступа, это означает, что она будет выполнена в конце работы программы в любом случае. А вот две предыдущие строчки будут выполнены, только если условие `if` окажется ложным.

Блоки кода в Python очень гибко устроены: внутри них можно писать любой другой код, в том числе условные операторы. Среди команд, которые выполняются, если условие `if` истинно («внутри `if`») или ложно («внутри `else`»), могут быть и другие условные операторы. Тогда команды, которые выполняются внутри этого внутреннего `if` или `else`, записываются с дополнительным отступом.

Изучите пример ниже. `elif` — это короткая запись для `else: if`. Если не пользоваться короткой записью, `if` пришлось бы писать на отдельной строчке и с отступом (а все, что внутри этого `if`, — с дополнительным отступом). Это не очень удобно, и `elif` избавляет от такой необходимости.

In [35]:
print('Представься, о незнакомец!')
name = input()
if name == 'Цезарь' or name == 'Caesar':
    print('Аве, Цезарь!')
    print('В честь какого бога устроим сегодня празднество?')
    god = input()
    if god == 'Юпитер':
        print('Ура Громовержцу!')
    # если оказалось, что имя бога не 'Юпитер', то проверяем, 
    # не равно ли оно строке 'Минерва'
    elif god == 'Минерва': 
        print('Ура мудрой воительнице!')
    # следующая строка будет выполнена, 
    # только если имя бога не 'Юпитер' и не 'Минерва'
    else: 
        print('Бога по имени', god, 'мы не знаем, но слово Цезаря - закон.')
    # эта команда будет выполнена независимо от того, 
    # какое имя бога ввёл пользователь, если только изначально
    # он представился Цезарем
    print('Слава императору!') 
else:
    print('Приветик.')
    print('Погода сегодня хорошая.')
print('Засим - заканчиваем.')

Представься, о незнакомец!
Аве, Цезарь!
В честь какого бога устроим сегодня празднество?
Ура мудрой воительнице!
Слава императору!
Засим - заканчиваем.


---
Условия if, elif, else и код в их блоках называются **ветвями** алгоритма. Сколько операторов if, некоторое количество elif, else есть в коде, столько у конструкции if-elif-else будет ветвей.

---

# Упражнения

У нас есть бизнес-план! Надо написать программу-тест, которая по некоторым простым вопросам выдаёт «строго индивидуальный» анализ личностных качеств.

Задайте пользователю два вопроса с возможными ответами «да» или «нет» (например, «Любите ли вы котиков?» и «Умеете ли вы программировать?») и считайте ответ. Если пользователь дал любой другой ответ на один из вопросов, сообщаем ему об ошибке и завершаем работу. Если же он ответил «да» или «нет» на каждый из вопросов, выдаём пользователю результат (например, «Вы обладаете незаурядным умом.»), причём для каждой возможной комбинации ответов результат должен быть уникальный.

In [39]:
# код
answer1 = input("Вы смотрите битву сильнейших? : (да/нет) ").strip().lower()
answer2 = input("Вы хотели бы иметь суперспобности? : (да/нет) ").strip().lower()
if answer1 == "да" or answer1 == "нет":
    if answer2 == "да" or answer2 == "нет":
        # результаты в зависимости от комбинации ответов
        if answer1 == "да" and answer2 == "да":
            print("Походу, многие люди о вас мало чего знают!")
        elif answer1 == "да" and answer2 == "нет":
            print("Вам нравится битва сильнейшних, но вы не хотите иметь суперспособности. Ничего страшного, вы и так хороши в другой сфере!")
        elif answer1 == "нет" and answer2 == "да":
            print("Вы не смотрите битву сильнейших, но вы бы хотели иметь суперспособности. Значит вы хотите стать супергером, например: Человеком-пауком!")
        else: #answer1 == "нет" and answer2 == "нет"
            print("Вы не смотрите битву сильнейших и не хотите иметь суперспособности. Зато вы классный человек!")
    else:
        print("Ошибка.")
else:
    print("Ошибка.")


Походу, многие люди о вас мало чего знают!


# Операции над строками

>Данный раздел приведен в целях освежить вашу память. Он уже встречался вам в предыдущем занятии, однако информация ниже пригодится для решения упражнений в конце этого блока. Поэтому **прочитайте** этот блок

Во всех примерах, которые мы рассматривали, переменные хранили строки. Мы вводили, выводили и хранили строки. Кроме уже описанных действий, строки еще можно складывать.

Давайте попробуем:

In [40]:
x = '10'
y = '20'
z = x + y
print(z)

1020


В данной программе мы задали переменным x и y значение, переменной z присвоили значение результата сложения x и y.

Результатом выполнения программы будет строка **1020**.

### И опять немного рекомендаций по оформлению (PEP 8): ставьте пробелы вокруг знаков операций (*, +, - и т. д.)

Правильно:

```python z = x + y```

~~Неправильно~~:

```python z=x+y```

## Конкатенация

Операция сложения для строк выполняет конкатенацию двух строк, то есть склеивает их содержимое вместе.

**Например**: операция «При» + «вет» в результате даст строку «Привет».

In [41]:
'при' + 'вет'

'привет'

Обратите внимание: запись: x + y = z недопустима. Оператор присваивания ожидает слева переменную, которой надо присвоить значение, а в правой части находится значение или выражение, которое надо сначала вычислить, а затем присвоить.

Мы могли сократить нашу программу и написать в таком виде:

In [42]:
x = '10'
y = '20'
print(x + y)

1020


Результат будет такой же, проверьте. Оператор ``print()`` сначала вычислил значение выражения x + y, а потом вывел на экран полученное значение. Однако если нам еще понадобится значение суммы x + y, то целесообразнее использовать первый вариант `` z = x + y``, чтобы не считать результат еще раз

## Дублирование

Для строк также можно выполнять умножение. Умножать можно строку на число или число на строку. Операция называется дублирование. В результате нее начальная строка будет повторена заданное количество раз.

Например: `3 * "20"` то же, что и `"20" * 3` и , результат будет `202020` и в том, и в другом случае.

Примеры использования:

In [43]:
'20' * 3

'202020'

In [44]:
x = '10'
y = '20'
print(x * 2 + y * 3)

1010202020


# Упражнения

Напишите программу, которая считывает с клавиатуры слово (или фразу, неважно), после чего выводит это же слово четыре раза на одной строке, не разделяя их пробелами.

In [48]:
# замените код тут
name=input("Введите слово или фразу: ")
print(name * 4)

ШарфШарфШарфШарф


# Команда in

Теперь рассмотрим новую команду для работы со строками — команду in.

Команда `in` позволяет проверить, что одна строка находится внутри другой.

Например: строка «на» находится внутри строки «сложная задача».

В таком случае обычно говорят, что одна строка является **подстрокой** для другой.


**Напоминаем, что с кодом, который реализует ветвление, необходимо попробовать разные комбинации ввода, чтобы лучше усвоить смысл конструкций ветвления**

In [53]:
text = input()
if 'хорош' in text and 'плох' not in text:
    print('Текст имеет положительную эмоциональную окраску.')
elif 'плох' in text and 'хорош' not in text:
    print('Текст имеет отрицательную эмоциональную окраску.')
else:
    print('Текст имеет нейтральную или смешанную эмоциональную окраску.')

Текст имеет нейтральную или смешанную эмоциональную окраску.


Первое условие окажется истинным, например, для строк «все хорошо» и «какой хороший день», но не для «ВсЕ ХоРоШо» и не для «что-то хорошо, а что-то и плохо». Аналогично второе условие окажется истинным для строк «все плохо», «плохое настроение» и т. д.

# Упражнения

Напишите программу, которая считывает одну строку, после чего выводит «МЯУ», если в введённой строке есть подстрока «кот», и «ГАВ» в противном случае.

>Для проверки наличия подстроки в другой строке, можно использовать оператор `in`. Ознакомьтесь с кодом ниже. Обратите внимание на то, что проверка происходит с учетом регистра букв (прописные, строчные). Это обуславливает разницу в выполнении двух ячеек ниже

In [54]:
text = 'кот вышел на улицу'
'кот' in text

True

In [55]:
text = 'Кот вышел на улицу'
'кот' in text

False

In [64]:
# тут код решения
# Считываем строку от пользователя
text = input("Введите строку: ")
if "кот" in text:
    print("МЯУ")  # Если подстрока найдена, выводим "МЯУ"
else:
    print("ГАВ")  # Если подстрока не найдена, выводим "ГАВ"


МЯУ


### Задача со звездочкой



Напишем программу, которая сумеет провести с пользователем короткий, более или менее осмысленный диалог.

Сначала программа спрашивает у пользователя, какое у него настроение. Затем ответ анализируется. Предполагается, что пользователь отвечает словом или фразой, записанной строчными буквами.

- Если в ответе пользователя есть ключевые слова вроде «хорошее», «прекрасно» (в любой форме), программа отвечает что-нибудь подходящее — например «Отлично, у меня тоже всё хорошо :)».
- Если же встречаются слова вроде «плохо», то нужно ответить что-то подходящее и на этот случай (например, «Ничего, скоро всё наладится»).
- Если ключевых слов того или другого типа не найдено, программа отвечает что-то нейтральное или извиняется и говорит, что не понимает пользователя. То же происходит, если в ответе есть «не» или вопросительный знак.

Случай, когда в ответе есть и «хорошие», и «плохие» слова, можно не рассматривать. 

In [70]:
# код
text = input("Как ваше настроение? ")
# Проверяем наличие ключевых слов в ответе
if "восхитительно" in text or "прекрасно" in text or "отлично" in text:
    print("Замечательно, я тоже чувствую себя хорошо.)")
elif "отвратительно" in text or "ужасно" in text or "не очень" in text:
    print("Не переживай, всё будет хорошо.")
elif "что" in text or "?" in text:
    print(" Я вас совсем не понимаю.")
else:
    print("Интересно, расскажите подробнее!")


Замечательно, я тоже чувствую себя хорошо.)
